### DISCUSSION TITLE
changeset cli for .NET world

### DISCUSSION THREAD

**Discussion by magiino** [changesets/changesets#1329] [Ideas]
Hello Changesets Team!

Me and my team would like to port Changesets tool to .NET world.

We need to manage automatic versioning, changelog generation and Nuget package releases in mono repo. We really like the principles which Changesets brings and we would like to bring it to .NET world!

The idea is to create Changesets as [dotnet tool](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools). Instead of **package.json** the tool would work with **.csproj**. Instead of **npm** repository the tool would publish nugets to **nuget** repository. The flow with generating changesets and CHANGELOG.md from changesets would stay the same.

The CLI would stay the same. The only difference would be, instead of calling npm
```
npm changeset add
npm changeset version
```
we would call dotnet
```
dotnet changeset add
dotnet changeset version
```
Ofc other commands would be also ported, they are just not mentioned here.

We would like to reuse your [GitHub release action](https://github.com/changesets/action).
Would you accept PR which would refactor current Action to work also with dotnet tool? We do not have any specific solution right now, we would like to know your thoughts about this.

If we will feel ready we would publish changeset dotnet tool here on GitHub as opensource project.

Thanks!

Marek

**Comment by mburumaxwell** [comment]
@magiino did you by any chance manage to start this effort?

**Reply by jvilimek** [reply]
Hi @mburumaxwell, we are in the starting phase of the project. Would you like to contribute somehow?

**Reply by mburumaxwell** [reply]
@jvilimek definitely interested for use with some monorepos I work in. What's the contribution plan? DM/Slack or just here?

**Reply by jvilimek** [reply]
the plan is to create a public repo and start project, create issues, discussion etc. there (if approved by our mgmt,legal etc).

**Reply by magiino** [reply]
Hello, we recently released the Changesets tool for the .NET world - `net-changesets`.

Check it out on GitHub:
👉 https://github.com/solarwinds/net-changesets

Feedback and contributions are welcome. 🚀

**Comment by jvilimek** [comment]
Update status: we are still open for cooperation! Recently @ThibautMarechal from [EVS](https://evs.com/) reached out for cooperation, thanks for that! Currently are starting this project, it's in the approval stage so hopefully, I would be able to share more soon!

**Reply by Andarist** [reply]
If you share some rough plan on what you plan to implement I could try to give more concrete thoughts about how I think/feel about it. Do you want to rewrite the whole Changesets CLI in dotnet and just keep the idea of Changesets as your "public API"?

**Reply by jvilimek** [reply]
> Do you want to rewrite the whole Changesets CLI in dotnet and just keep the idea of Changesets as your "public API"?

yeah, that's roughly how we plan to do it. Ideally re-use what could be re-used, i.e. the github action.

**Reply by magiino** [reply]
Yes, as you mentioned, the Changesets CLI "public" API remains the same, and we are keeping the original idea behind Changesets. Currently, the CLI API does not support all parameters, which is fine since we don't need more at the moment. The only difference is in config.json, but that’s acceptable because it's behind the "public" API. We’re also working with .csproj files and a NuGet repository instead of project.json and an npm repository.
 
Please let us know your thoughts!

**Comment by BoatPartyJesus** [comment]
Any update on this? We're interested in a dotnet solution, keen to see if we can help

**Reply by jvilimek** [reply]
Sorry, it took quite a while. But finally, we are public, and hopefully things will accelerate now!

**Comment by risenxxx** [comment]
Currently I'm creating my own solution very similar to changesets

**Reply by magiino** [reply]
Hello, we recently released the Changesets tool for the .NET world - `net-changesets`.

Check it out on GitHub:
👉 https://github.com/solarwinds/net-changesets

Feedback and contributions are welcome. 🚀

**Comment by jan-bezdicek-swi** [comment]
Hello everyone,

I lead the team at SolarWinds developing a .NET Changeset tool. We **intend to release it as an open-source project** soon, but there’s no confirmed date yet, and our priorities may shift over time. We truly appreciate your interest and will share updates as the project progresses.

Thank you, 
Jan, SolarWinds

**Comment by Mathieu-Lichtsteiner** [comment]
Hi everyone,

I just stumbled over this discussion, because we are currently using changesets for dotnet through package.json files and automated <AssemblyVersion>, <Version>, ... syncs. Since the whole PackageReference and ProjectReference system is built a bit different than in JavaScript I would propose to go in the direction of implementing an adapter for msbuild files (maybe more generic: XML).
If I were to put some effort into this, how would the possibility for integration look like? Is there any interest in opening the architecture and or including different dependency management systems?
(To be upfront, I haven't really looked at any code yet, I'm sorry...)

Best regards,
Mathieu

**Reply by jvilimek** [reply]
Hi @Mathieu-Lichtsteiner,
not sure, if still relevant, but finally, the project has been published https://github.com/solarwinds/net-changesets
Currently, we support the CSPROJ files instead of package.json. But the rest should be pretty similar to `changeset` npm tool

**Comment by JanBez** [comment]
Hi everyone, following up on this thread. We have just open-sourced **.NET Changesets**, a .NET CLI that brings the Changesets workflow to .NET and multi-package repos. It handles semantic versioning, generates a unified `CHANGELOG.md`, respects dependency graph bumps, and can publish packages to a NuGet feed.

Repo: [https://github.com/solarwinds/net-changesets](https://github.com/solarwinds/net-changesets) 
NuGet: [https://www.nuget.org/packages/SolarWinds.Changesets](https://www.nuget.org/packages/SolarWinds.Changesets)

**Install**

```bash
dotnet tool install solarwinds.changesets --global
```

Docs, roadmap, and contribution guidelines are in the repo. The project is Apache-2.0 licensed. We would love feedback and contributors, especially from folks who asked for a .NET port here. Thanks to the Changesets maintainers for the original workflow and inspiration. 

---
Jan,
[SolarWinds](http://www.solarwinds.com)

**Comment by alexaka1** [comment]
Fwiw, I use changesets with exclusively dotnet projects. What I have started doing most recently is create a `Version.props` file next to the package.json, that has the `<Version>` property with the same value as the package.json, and is referenced in my `csproj`. I have a custom bash script that is invoked by the changeset/action that first invokes the `changeset version` command, and then just uses jq and xmlint to get the version and update the `props` file. Since this is done in the same step within the action, this update persists in the PR opened by the action. I would recommend using simply `sed` instead of `xmlint` as it is abandoned, and if you can control that your `props` file won't contain any crazy formatting, then sed is more than enough to replace the version.

**Reply by jvilimek** [reply]
Thanks @alexaka1 for the comment!
This is indeed one of the options we discussed internally. In the end, we did not go that way, as
- it won't solve project dependencies. In our current solution, for [version](https://github.com/solarwinds/net-changesets/blob/main/docs/commands-implementation-details.md#version) command, we change (patch) version also for all projects that **depend** on the project being handled. The documentation is not ideal, though, we need to improve it so it's clearer
- `version.props` and `package.json` are additional 2 files that you need to have in each project directory. And you need to keep it in sync. And implement special (CI/build) logic for that.
- You wrote that you have a special script as part of your `changeset version` CI/build pipeline. So, if a developer runs this manually, the script is not executed, and the files are out of sync, right? Also, if you edit the files directly, it can break the logic, right?


UPDATE: Additionally, one concern with the original changeset was that .NET developers would need to install `node.js` / `npm` solely for the purpose of creating changesets/versioning.

**Reply by alexaka1** [reply]
> - it won't solve project dependencies

What do you mean by that? I just link the same Version.props file to all of my projects that belong to the same app. This way I avoid spamming tags for `Foo.Utils.Shared` or whatever, but all projects share the same version number from the version.props.

> version.props and package.json are additional 2 files that you need to have in each project directory. And you need to keep it in sync. And implement special (CI/build) logic for that.

Yes, but I have already set this up, so it is zero cost after that. I don't see any issues in every packagejson having a sibling props file, as changesets requires package.json anyway, so that's also a _special file_ you need to remember to not edit manually.

> You wrote that you have a special script as part of your changeset version CI/build pipeline.

I have just finished this in one of my oss projects that I can share with you. I have opted to do fully automated releases, and there is a ruleset that prohibits tag creation in the repo that match the tag names of what changeset would create. Only the release bot is allowed to create/update/delete these tags.

https://github.com/alexaka1/distroless-dotnet-healthchecks/blob/9bd7935b4a10c8846a62c964f9b4a9ccae48397c/.github/workflows/version.yml

**Reply by jvilimek** [reply]
Thanks for sharing your code!

Let me elaborate on the `project dependencies`. We probably have a slightly different use case than you. Among other repositories, we have a `mono-repo` repository that contains many projects producing *nuget* libraries. These have dependencies on each other. So we have `<ProjectReference .../>` in CSPROJ files. And we want to increase the version of all affected libraries, if any of their dependencies have changed. 

Let's have `Lib1` and `Lib2` libraries (nugets/project), where `Lib1` -depends-on-> `Lib2`, both in the same repository. When I use `changeset` to add **minor** change to `Lib2`, when calling the `version` command, I want the `Lib2` version **minor** part to be increased, and **automatically** the `Lib1` version **patch** part to be increased as well (with a changelog like `dependency Lib2 bumped to version X.Y`).

Hope it's more clear now.

**Reply by alexaka1** [reply]
Isn't that just the fixed/linked package feature? https://github.com/changesets/changesets/blob/main/docs/fixed-packages.md

I have used `fixed` for a short time with a .net project. Same principle, a 2 package.jsons for the two csproj files.
